### PR TITLE
Remove some parser S/R conflicts

### PR DESCRIPTION
**Pull request by fbs**
This reduces the amount of conflicts from:

```
/vagrant/src/parser.yy: warning: 126 shift/reduce conflicts [-Wconflicts-sr]
/vagrant/src/parser.yy: warning: 31 reduce/reduce conflicts [-Wconflicts-rr]
```

To:

```
/vagrant/src/parser.yy: warning: 28 shift/reduce conflicts [-Wconflicts-sr]
/vagrant/src/parser.yy: warning: 29 reduce/reduce conflicts [-Wconflicts-rr]
```

There is a lot of refactoring that has to happen if we want to get rid of all of them eventually. The last 8 statements in `expr` are the main cause of all the issues and will be tricky to get right. My goal is to get rid of all the R/R and most of the S/R so that we can alert when new cases show up instead of having them silently sneak in.

### CODE REVIEW

File: src/parser.yy [hunk 1]
```diff
@@ -316,6 +325,10 @@ map : MAP               { $$ = new ast::Map($1, @$); }
 var : VAR { $$ = new ast::Variable($1, @$); }
     ;
 
+variable : var { $$ = $1; }
```
Comment (ajor) [inline src/parser.yy]: Maybe something like `var_or_map` instead? The name "variable" confused me when I first read through, thinking it didn't include maps.

Comment (fbs) [reply src/parser.yy]: My issue with `map_or_var` is that once a 3rd type of variable gets added all those cases have to be renamed. But I guess that is unlikely to happen. I'll rename it

**Comment by fbs** [comment]
Updated @ajor

**Comment by ajor** [comment]
I've merged this into master manually to fix the conflicts. Commits:
b3680e6b47354dd719fe9a820a293f8d87641a22
c126441f56f9b1c01b45a83e2987311b7e57f4c0

**Pull request closed without merging**